pkg/controllers/scheduling/predicate.go

```diff
@@ -0,0 +1,86 @@
+package scheduling
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	clusterapiv1 "github.com/open-cluster-management/api/cluster/v1"
+	clusterapiv1alpha1 "github.com/open-cluster-management/api/cluster/v1alpha1"
+)
+
+// matchWithClusterPredicates returns a slice of clusters. Each of them must match at least one of the predicates.
+func matchWithClusterPredicates(predicates []clusterapiv1alpha1.ClusterPredicate, clusters []*clusterapiv1.ManagedCluster) ([]*clusterapiv1.ManagedCluster, error) {
+	if len(predicates) == 0 {
+		return clusters, nil
+	}
+	if len(clusters) == 0 {
+		return clusters, nil
+	}
+
+	// prebuild label/claim selectors for each predicate
+	predicateSelectors := [][]labels.Selector{}
```

Comment:
nit: rather than a 2d array, I prefer an array of structs

```go
type predicateSelector struct{
  labelSelector
  claimSelector
}
```

Reply:
Code refactored as suggested.